### PR TITLE
Build metaobject without Qt private headers dependency

### LINKED ISSUE #19: Remove Qt private header dependency
Currently libqmlbind depends on Qt private headers:

> This project is using private headers and will therefore be tied to this specific Qt module build version. Running this project against other versions of the Qt modules may crash at any arbitrary point. This is not a bug, but a result of using Qt internals. You have been warned!

To make libqmlbind more portable,  Qt private header dependency should be removed.

### PR DESCRIPTION
**Pull request by seanchas116**
To resolve #19 

* Setup metaobject data inside `MetaObject` class
* Refactor API
  * Remove `qmlbind_metaobject` from public API
  * Rename `qmlbind_exporter` to `qmlbind_metaclass`
* Remove dependency on Qt private headers

### CODE REVIEW

**Comment by florianjacob** [comment]
The general direction really seems the way to go, not only to stop us from depening on the private headers, but also to stop us from some undocumented implementation hidden somewhere in the details and private classes of Qt. :+1: 

Would you prefer to merge this before or after #15?

**Comment by seanchas116** [comment]
I'd prefer the refactored code so I'd like to merge after #15 :smiley:

**Comment by seanchas116** [comment]
I was able to get rid of dependency on Qt private headers and I'm doing further refactoring of `qmlbind_exporter` and `qmlbind_metaobject`. Now `qmlbind_metaobject` builds its metaobject data by itself and `qmlbind_exporter` just holds methods/properties information and client callbacks and doesn't "export" anything.
I'm thinking of renaming `qmlbind_exporter` to `qmlbind_classinfo` or something. @florianjacob and @rainbyte , do you have any idea of the new name or better refactoring idea?

**Comment by florianjacob** [comment]
I really like how `MetaObject` does not depend anymore on `Exporter` and need to call it to get various stuff. :sparkles: 

:+1: in favour of renaming `qmlbind_exporter`, but I it is quite hard to think of good names for it!

Maybe `qmlbind_metaclass` would be an alternative. I tried out how those names would feel:
```
qmlbind_classinfo *classinfo = qmlbind_classinfo_new();
qmlbind_classinfo_add_method(classinfo, "hello", 2);
qmlbind_classinfo_add_property(classinfo, "value", "valueChanged");
qmlbind_metaobject *metaobject = qmlbind_metaobject_new(classinfo);
qmlbind_classinfo_release(classinfo);
```

```
qmlbind_metaclass *metaclass = qmlbind_metaclass_new();
qmlbind_metaclass_add_method(metaclass, "hello", 2);
qmlbind_metaclass_add_property(metaclass, "value", "valueChanged");
qmlbind_metaobject *metaobject = qmlbind_metaobject_new(metaclass);
qmlbind_metaclass_release(metaclass);
```

I like how `qmlbind_metaclass_add_method(metaclass, "hello", 2);` quite literally says “add this method to the metaclass.“, and because a metaclass object is an object that it describes the class, so the class generated later on will have that method. The downside of it is that all that `meta` in `qmlbind_metaobject` and then `qmlbind_metaclass` could get a little confusing, though.

**Comment by seanchas116** [comment]
Another idea I came up with is removing `qmlbind_metaobject` from public API. Instead we can create metaobject internally on `qmlbind_engine_new_wrapper` and `qmlbind_register_type`.
Then we can rename `qmlbind_exporter` to `qmlbind_metaclass`, and the API will become cleaner.

**Comment by seanchas116** [comment]
I finished renaming `qmlbind_exporter` to `qmlbind_metaclass` and removing `qmlbind_metaobject`.
@florianjacob can you review the pull request?

File: qmlbind/include/qmlbind/metaclass.h [hunk 1]
```diff
@@ -6,56 +6,54 @@ extern "C" {
...
  * \param classObject will be given as parameter to the `new_object` function of the provided
  * `qmlbind_client_callbacks`. You can use it to hand over the equivalent of class objects in your language to
  * implement a generic `new_object` function. If you don't need it, just hand over a nullpointer. This function does not
  * take ownership of the classObject, but it still needs to be valid for as long as new objects are created (i.e. the
  * whole application lifetime).
  *
- * \param className is used for registering the metaobject created with this exporter in the metaobject system.
+ * \param className is used for registering the metaobject created with this metaclass in the metaobject system.
  *
  * \param interfaceHandlers is used to create, use and delete instances of the metaobject to build.
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/metaclass.h]: Seems like I forgot to rename `\param interfaceHandlers` to `\param callbacks` here in my `qmlbind_interface_handlers` rename. :laughing: Maybe you could fix that here, too?

File: qmlbind/include/qmlbind/metaclass.h [hunk 2]
```diff
@@ -6,56 +6,54 @@ extern "C" {
...
+ * \brief Creates a new `qmlbind_metaclass`.
  *
  * \param classObject will be given as parameter to the `new_object` function of the provided
  * `qmlbind_client_callbacks`. You can use it to hand over the equivalent of class objects in your language to
  * implement a generic `new_object` function. If you don't need it, just hand over a nullpointer. This function does not
  * take ownership of the classObject, but it still needs to be valid for as long as new objects are created (i.e. the
  * whole application lifetime).
  *
- * \param className is used for registering the metaobject created with this exporter in the metaobject system.
+ * \param className is used for registering the metaobject created with this metaclass in the metaobject system.
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/metaclass.h]: Because the user now registers metaclasses instead of -objects via `qmlbind_metaclass_register()`, I think it makes sense to rephrase it to something like this: `\param className is used for registering this metaclass to the metaobject system.`

File: qmlbind/src/metaobject.cpp [hunk 3]
```diff
@@ -98,4 +106,165 @@ int MetaObject::metaCall(QObject *object, Call call, int index, void **argv) con
...
+    }
+
+    // properties
+    propertyOffsetPlaceholder.setCurrentPos();
+    for (const auto &prop : mProperties) {
+        metadata << strData.add(prop.name) << jsValueType;
+        auto signalIndex = mSignalIndexMap.value(prop.notifySignalName, -1);
+        if (signalIndex == -1) {
+            qWarning() << "qmlbind: no such notify signal" << prop.notifySignalName << "in class" << className;
+            metadata << 0x0009510b;
```
Comment (florianjacob) [inline qmlbind/src/metaobject.cpp]: You could add a comment here about what those two magic hex numbers in this if/else here mean.

File: qmlbind/include/qmlbind/metaclass.h [hunk 4]
```diff
@@ -98,17 +96,39 @@ QMLBIND_API void qmlbind_exporter_add_signal(
...
+ * import <uri> <versionMajor>.<versionMinor>
+ * ```
+ *
+ * After registering, metaclass changes such as method addition will not be reflected in the QML type system.
+ *
+ * This is libqmlbind's rough equivalent of
+ * [QQmlEngine::registerType](https://doc.qt.io/qt-5/qqmlengine.html#qmlRegisterType).
+ */
+QMLBIND_API int qmlbind_metaclass_register(
+    qmlbind_metaclass *self,
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/metaclass.h]: This should be `const qmlbind_metaclass *self` because `createMetaObject()` is also `const`. It then also implies the fact that ownership of the metaclass is not transfered and it can be used again by the client, e.g. for `qmlbind_engine_new_wrapper()`.

File: qmlbind/include/qmlbind/metaclass.h [hunk 5]
```diff
@@ -98,17 +96,39 @@ QMLBIND_API void qmlbind_exporter_add_signal(
...
+ *
+ * `uri`, `versionMajor` and `versionMinor` is the package uri and version under which this `metaclass` is made
+ * available to QML. After calling this function, you can import the component as:
+ *
+ * ```
+ * import <uri> <versionMajor>.<versionMinor>
+ * ```
+ *
+ * After registering, metaclass changes such as method addition will not be reflected in the QML type system.
+ *
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/metaclass.h]: Let's add a note here about the required lifetime of the passed metaclass? As I see it, a client could destoy the metaclass directly after registering if he wants to, but should keep it around in case he wants to use `qmlbind_engine_new_wrapper` later on.

File: qmlbind/include/qmlbind/engine.h [hunk 6]
```diff
@@ -121,12 +121,14 @@ QMLBIND_API qmlbind_value *qmlbind_engine_new_array(qmlbind_engine *self, int le
  *
  * If `object` is a null pointer, this function returns a null value.
  *
+ * After creating a wrapper, metaclass changes such as method addition will not be reflected in the created wrapper.
+ *
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/engine.h]: Let's add a note here about the required lifetime of the passed metaclass, too. Again, a client could destoy the metaclass directly after creating the wrapper, if he won't create any new wrappers anymore.

The client will probably implement a function that generates metaclasses for him, so he could plan to destroy each metaclass directly after use and create a new one later on, even if it's identical (has the same className, methods, properties etc.) I'm not sure whether this will lead to internal blow-up or strange side effects, but maybe we should still discourage this, and require that all metaclasses passed to `_new_wrapper` and `_metaclass_register` that have the same name have to be the identical metaclass pointer.

I find it really hard to express this in an understandable way, though. :sweat:

**Comment by florianjacob** [comment]
General feedback: I think it's a brilliant solution to remove `qmlbind_metaobject` from the public API, いけがみ さま! It really was not neccessary to have both types anymore. I could have noticed this earlier because of the fact that `qmlbind_metaobject` did not provide any meaningful methods for the client on its own, but only existed to be passed to libqmlbind for internal use, but the idea just not came into my head. I really like it! :rocket: 

:+1: for considering what happens if the client modifies the `qmlbind_metaclass` afterwards.

:+1: for caching the created internal metaobject via a shared_ptr.

I only found some documentation issues, see my other comments for that.

**Comment by seanchas116** [comment]
I fixed the docs. Thanks all for suggestions and :+1: s!

File: qmlbind/include/qmlbind/metaclass.h [hunk 7]
```diff
@@ -14,6 +14,9 @@ extern "C" {
  *
  * Internally, the `qmlbind_metaclass` creates a [QMetaObject](https://doc.qt.io/qt-5/qmetaobject.html) for the wrapper Qt objects.
  *
+ * It is not recommended to use multiple `qmlbind_metaclass` instances for the same class name
+ * to prevent possible internal blow-up or strange side effects.
```
Comment (florianjacob) [inline qmlbind/include/qmlbind/metaclass.h]: Right to the point! :boom:

File: qmlbind/src/metaobject.cpp [hunk 8]
```diff
@@ -6,6 +6,25 @@
 
 namespace QmlBind {
 
+// See qmetaobject_p.h for flag details
```
Comment (florianjacob) [inline qmlbind/src/metaobject.cpp]: Today it came to my mind that a hint where to look all this up would be nice as it's obviously not in the Qt documentation, but you already added that in the meantime, nice. :smile:

File: qmlbind/src/metaobject.cpp [hunk 9]
```diff
@@ -6,6 +6,25 @@
...
+
+enum PropertyFlags {
```
Comment (florianjacob) [inline qmlbind/src/metaobject.cpp]: This was more than I excepcted that could be done here, this makes me able to reason nicely about the code even without learning the details of `qobject_p.h`. :+1:

**Comment by florianjacob** [comment]
Let's get this merged! :shipit:
